python/lsst/display/firefly/footprints.py

```diff
@@ -0,0 +1,202 @@
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+##
+# @file
+# @brief Support for LSST footprints, specific to Firefly
+
+from copy import deepcopy
+import io
+import tempfile
+
+import numpy as np
+from astropy.io.votable.tree import VOTableFile, Resource, Field, Info
+from astropy.io.votable import from_table
+from astropy.table import Table
+import astropy.units as u
+
+import lsst.afw.geom as afwGeom
+
+
+
+def record_selector(record, selection):
+    """Select records from source catalog
+
+    Parameters:
+    -----------
+    record : `lsst.afw.detect.SourceRecord`
+        record to select
+    selection : `str`
+        'all' to select all records. 'blended parents' to select records with
+        more than zero children. 'deblended children' to select records with
+        non-zero parents. 'isolated' to select records that are not blended,
+        meaning zero parents and zero children.
+        Values to check for sel
+    """
+    nChildren = record.get('deblend_nChild')
+    parentId = record.getParent()
+    if selection == 'all':
+        return True
+    elif selection == 'blended parents':
+        return (nChildren > 0)
+    elif selection == 'deblended children':
+        return (parentId > 0)
+    elif selection == 'isolated':
+        return ((parentId == 0) and (nChildren == 0))
+    else:
+        raise RuntimeError('invalid selection: {}'.format(selection) + 
+                           '\nMust be one of "all", "blended parents", ' +
+                           '"deblended children", "isolated"')
+
+
+def create_footprints_table(catalog, xy0 = afwGeom.Point2I(0,0),
+                            insert_column=6):
+    """make a VOTable of SourceData table and footprints
+
+    Parameters:
+    -----------
+    catalog : `lsst.afw.table.SourceCatalog`
+            Source catalog from which to display footprints.
+    xy0 : `lsst.afw.geom.Point2I`, default afwGeom.Point2I(0,0)
+        Pixel origin to subtract off from the footprint coordinates.
+        Normally this can be left at the default.
+    insert_column : `int`
+        Column at which to insert the "family_id" and "category" columns
+
+    Returns:
+    --------
+    `astropy.io.votable.VOTableFile`
+        VOTable object to upload to Firefly
+    """
+    with tempfile.NamedTemporaryFile() as fd:
+        catalog.writeFits(fd.name)
+        sourceTable = Table.read(fd.name, hdu=1)
+
+    # Fix invalid unit strings of "seconds"
+    for col_name in ['modelfit_CModel_dev_time', 'modelfit_CModel_initial_time',
+                     'modelfit_CModel_exp_time']:
+        if col_name in sourceTable.colnames:
+            sourceTable[col_name].unit = u.s
+
+    inputvofile = from_table(sourceTable)
+    inputvotable = inputvofile.get_first_table()
+    outtable = deepcopy(inputvotable)
```

Comment:
Comments would be welcome here - it looks like you're converting a `votable` into a `votable`, extracting a `votable` from that, and then deep-copying that into a `votable`.

Reply:
Comments added. Adding columns to the output table causes the existing table to be destroyed, so the copy is needed so the input records can be copied to the output.